### PR TITLE
DDT-760: Location logic

### PR DESCRIPTION
**Pull request by mp-assco**

### CODE REVIEW

**Comment by douradofrazer** [comment]
I'd suggest moving the pom.xml changes to dcsa-core.

File: src/main/java/org/dcsa/core/events/service/impl/LocationServiceImpl.java [hunk 1]
```diff
@@ -14,68 +15,180 @@
...
+                .map(Optional::of)
+                .defaultIfEmpty(Optional.empty()),
+            facilityService
+                .findByIdOrEmpty(location.getFacilityID())
+                .map(Optional::of)
+                .defaultIfEmpty(Optional.empty()))
+        .flatMap(
+            t2 -> {
+              LocationTO locTO = locationMapper.locationToDTO(location);
+              t2.getT1().ifPresent(locTO::setAddress);
```
Comment (pimg) [inline src/main/java/org/dcsa/core/events/service/impl/LocationServiceImpl.java:189]: setters can be moved to the mapper

File: src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java [hunk 2]
```diff
@@ -16,32 +16,48 @@
...
-                    Objects.requireNonNull(facilityCode, "facilityCode")
-                ).switchIfEmpty(Mono.error(new CreateException("Cannot find any facility with UNLocationCode + Facility code: "
-                        + unLocationCode + ", " + facilityCode)));
+  @Override
+  public Mono<Facility> findByUNLocationCodeAndFacilityCode(
+      String unLocationCode,
+      FacilityCodeListProvider facilityCodeListProvider,
+      String facilityCode) {
+    BiFunction<String, String, Mono<Facility>> method;
+    switch (Objects.requireNonNull(facilityCodeListProvider, "facilityCodeListProvider")) {
```
Comment (pimg) [inline src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java:35]: "facilityCodeListProvider" seems insufficient for a error message

File: src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java [hunk 3]
```diff
@@ -16,32 +16,48 @@
...
+      case BIC:
+        method = facilityRepository::findByUnLocationCodeAndFacilityBICCode;
+        break;
+      default:
+        throw new CreateException(
+            "Unsupported facility code list provider: " + facilityCodeListProvider);
     }
+    return method
+        .apply(
+            Objects.requireNonNull(unLocationCode, "unLocationCode"),
```
Comment (pimg) [inline src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java:48]: "unLocationCode" seems insufficient for an error message

File: src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java [hunk 4]
```diff
@@ -16,32 +16,48 @@
...
+            Objects.requireNonNull(facilityCode, "facilityCode"))
```
Comment (pimg) [inline src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java:49]: "facilityCode" seems insufficient for an error message

File: src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java [hunk 5]
```diff
@@ -16,32 +16,48 @@
...
+    BiFunction<String, String, Mono<Facility>> method;
+    switch (Objects.requireNonNull(facilityCodeListProvider, "facilityCodeListProvider")) {
+      case SMDG:
+        method = facilityRepository::findByUnLocationCodeAndFacilitySMDGCode;
+        break;
+      case BIC:
+        method = facilityRepository::findByUnLocationCodeAndFacilityBICCode;
+        break;
+      default:
+        throw new CreateException(
```
Comment (pimg) [inline src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java:43]: this method is deprecated better to use the new alternative

File: src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java [hunk 6]
```diff
@@ -16,32 +16,48 @@
...
-                    Objects.requireNonNull(unLocationCode, "unLocationCode"),
-                    Objects.requireNonNull(facilityCode, "facilityCode")
-                ).switchIfEmpty(Mono.error(new CreateException("Cannot find any facility with UNLocationCode + Facility code: "
-                        + unLocationCode + ", " + facilityCode)));
+  @Override
+  public Mono<Facility> findByUNLocationCodeAndFacilityCode(
+      String unLocationCode,
+      FacilityCodeListProvider facilityCodeListProvider,
+      String facilityCode) {
+    BiFunction<String, String, Mono<Facility>> method;
```
Comment (pimg) [inline src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java:34]: consider changing switch statement when only two cases are supported:
`Mono.just(facilityCodeListProvider).filter(facilityCodeListProvider1 -> facilityCodeListProvider1 == FacilityCodeListProvider.SMDG)
      .flatMap(smdgFacility -> facilityRepository.findByUnLocationCodeAndFacilitySMDGCode(unLocationCode, facilityCode))
      .switchIfEmpty(facilityRepository.findByUnLocationCodeAndFacilityBICCode(unLocationCode, facilityCode));`

**Comment by pimg** [comment]
Missing tests for FacilityService

File: src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java (hunk 5 above)
Comment (mp-assco) [reply src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java]: Fixed

File: src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java (hunk 4 above)
Comment (mp-assco) [reply src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java]: Updated

File: src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java (hunk 3 above)
Comment (mp-assco) [reply src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java]: Updated

File: src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java (hunk 2 above)
Comment (mp-assco) [reply src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java]: Updated

File: src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java (hunk 6 above)
Comment (nt-as) [reply src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java:36]: > consider changing switch statement when only two cases are supported: `Mono.just(facilityCodeListProvider).filter(facilityCodeListProvider1 -> facilityCodeListProvider1 == FacilityCodeListProvider.SMDG) .flatMap(smdgFacility -> facilityRepository.findByUnLocationCodeAndFacilitySMDGCode(unLocationCode, facilityCode)) .switchIfEmpty(facilityRepository.findByUnLocationCodeAndFacilityBICCode(unLocationCode, facilityCode));`

Comment (mp-assco) [reply src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java:36]: Would this befine?
```java
    return Mono.just(facilityCodeListProvider)
        .flatMap(
            facilityCodeListProvider1 -> {
              if (facilityCodeListProvider1 == FacilityCodeListProvider.SMDG) {
                return facilityRepository.findByUnLocationCodeAndFacilitySMDGCode(
                    unLocationCode, facilityCode);
              } else {
                return facilityRepository.findByUnLocationCodeAndFacilityBICCode(
                    unLocationCode, facilityCode);
              }
            })
        .switchIfEmpty(
            Mono.error(
                ConcreteRequestErrorMessageException.invalidParameter(
                    "Cannot find any facility with UNLocationCode + Facility code: "
                        + unLocationCode
                        + ", "
                        + facilityCode)));
```

Comment (nt-as) [reply src/main/java/org/dcsa/core/events/service/impl/FacilityServiceImpl.java:36]: That would work for me as well.

**Review by pimg — APPROVED** [review]

**Pull request merged**